src/main/java/org/osiam/storage/query/UserFilterExpression.java

```diff
@@ -0,0 +1,93 @@
+package org.osiam.storage.query;
+
+import org.osiam.resources.scim.User;
+import org.osiam.storage.entities.UserEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Locale;
+
+public class UserFilterExpression extends FilterExpression<UserEntity> {
+
+    private Field field;
+
+    UserFilterExpression(String field, FilterConstraint constraint, String value) {
+        super(constraint, value);
+        this.field = new Field(field);
+    }
+
+    @Override
+    public UserFilterExpression.Field getField() {
+        return field;
+    }
+
+    public static class Field implements FilterExpression.Field<UserEntity> {
+        private static final Logger logger = LoggerFactory.getLogger(UserFilterExpression.Field.class);
+        private String urn;
+        private String name;
+        private UserQueryField queryField;
+
+        public Field(String field) {
+            if (field.toLowerCase().startsWith(User.SCHEMA.toLowerCase())) {
+                if (field.charAt(User.SCHEMA.length()) == '.') {
+                    throw new IllegalArgumentException(String.format("Period (.) is not a valid field separator: %s", field));
+                }
+                name = field.substring(User.SCHEMA.length() + 1).toLowerCase(Locale.ENGLISH);
+            } else {
+                name = field.toLowerCase(Locale.ENGLISH);
+            }
+
+            // Try to resolve a core schema field.
+            queryField = UserQueryField.fromString(name);
+
+            urn = User.SCHEMA;
+            // we can't recognize a core schema field, it's an extension.
+            if (queryField == null) {
+
+                // lastIndexOf returns -1 if char can not be found in the string
```

Comment:
By documenting the algorithm I rather meant describing the high level thing we are doing here (measuring the length of the last parts separated by "." or ":") and why this works (no ":" and "." are allowed in attribute names), you know?